docs: add instructions on using repo finder (#350)
oracle/macaron: docs/source/pages/using.rst

- ``find_repos`` (Values: True or False) - Enables or disables the Repository Finding feature.
- ``artifact_repositories`` (Values: List of URLs) - Determines the remote artifact repositories to attempt to retrieve dependency information from.
- ``repo_pom_paths`` (Values: List of POM tags) - Determines where to search for repository information in the POM files. E.g. scm.url.
- ``find_parents`` (Values: True or False) - When enabled, the Repository Finding feature will also search for repository URLs in parents POM files of the current dependency.
- ``artifact_ignore_list`` (Values: List of GAs) - The Repository Finding feature will skip any artifact in this list. Format is "GroupId":"ArtifactId". E.g. org.apache.maven:maven

.. note:: Finding repositories requires at least one remote call, adding some additional overhead to an analysis run.

-------------------------------------
Analyzing a locally cloned repository
-------------------------------------

If you have a local repository that you want to analyze, Macaron also supports running the analysis against a local repository.

> ''''''''''''''''''

Reviewer: Looks like this line is added accidentally and is causing an error: https://github.com/oracle/macaron/actions/runs/5450632046/jobs/9916079385?pr=350#step:8:9763

- `micronaut-core.html <../_static/examples/micronaut-projects/micronaut-core/analyze_with_sbom/micronaut-core.html>`__
- `micronaut-core.json <../_static/examples/micronaut-projects/micronaut-core/analyze_with_sbom/micronaut-core.json>`__

'''''''''''''''''''''''''''
Analyzing more dependencies
'''''''''''''''''''''''''''

In some cases the dependencies that Macaron discovers lack a direct connection to a repository for it to analyze. To improve results in these instances, the Repository Finding feature can be enabled. This feature makes use of a dependency's identifying information that can be found using the package registries located on the Internet.

.. note:: The Repository Finding feature currently only works for Java projects via SCM meta data found within artifact POM files.

> This feature is enabled by default. To disable, or configure its behaviour in other ways, a custom ``defaults.ini`` should be passed to Macaron during execution. Under the ``repofinder.java`` header, five options exist: ``find_repos``, ``artifact_repositories``, ``repo_pom_paths``, ``find_parents``, ``artifact_ignore_list``. These options behave as follows:

Reviewer: It would be helpful to add a reference like this:

```
See :ref:`dump-defaults <reference to docs/source/pages/cli_usage/action_dump_defaults.rst >`, the CLI command to dump the default configurations in ``defaults.ini``. After making changes, see :ref:`analyze <analyze-action-cli>` CLI command for the option to pass the modified ``defaults.ini`` file.
```

You need to add a reference in `docs/source/pages/cli_usage/action_dump_defaults.rst` but `analyze-action-cli` reference already exists.

Author: I've tried to. Not sure if it's displaying as it should be though.

- ``find_repos`` (Values: True or False) - Enables or disables the Repository Finding feature.
- ``artifact_repositories`` (Values: List of URLs) - Determines the remote artifact repositories to attempt to retrieve dependency information from.
- ``repo_pom_paths`` (Values: List of POM tags) - Determines where to search for repository information in the POM files. E.g. scm.url.
- ``find_parents`` (Values: True or False) - When enabled, the Repository Finding feature will also search for repository URLs in parents POM files of the current dependency.
- ``artifact_ignore_list`` (Values: List of GAs) - The Repository Finding feature will skip any artifact in this list. Format is "GroupId":"ArtifactId". E.g. org.apache.maven:maven

.. note:: Finding repositories requires at least one remote call, adding some additional overhead to an analysis run.

> 

Reviewer 2: I think we should add an example snippet of the custom `defaults.ini` which enable this feature so that it's clearer for the user.

requirement. For example, `the Mysql Connector/J <https://slsa.dev/spec/v0.1/requirements#build-service>`_ dependency in
the Micronaut MuShop project does not pass the ``build_service`` check, but can be manually investigated and exempted if trusted. Overall, policies expressed in Datalog can be
enforced by Macaron as part of your CI/CD pipeline to detect regressions or unexpected behavior.

---------------------------
Analyzing more dependencies
---------------------------

In some cases the dependencies that Macaron discovers lack a direct connection to a repository for it to analyze. To improve results in these instances, the Repository Finding feature can be enabled. This feature makes use of a dependency's identifying information and the matching information that can be found within package managing repositories located on the Internet.

.. note:: The Repository Finding feature currently only works for Java projects via SCM meta data found within artifact POM files.

> Enabling this feature requires modification of the ``defaults.ini`` config file located at: ``<macaron_root_directory>/src/config/``. Under the ``repofinder.java`` header, four options of importance exist: ``find_repos``, ``artifact_repositories``, ``repo_pom_paths``, ``find_parents``. These options behave as follows:

Reviewer: Can you please clarify that by default repo finder is enabled?

Reviewer: It might be a good idea to hint at the potential performance overhead, and that users can disable it if they want to.

> Enabling this feature requires modification of the ``defaults.ini`` config file located at: ``<macaron_root_directory>/src/config/``. Under the ``repofinder.java`` header, four options of importance exist: ``find_repos``, ``artifact_repositories``, ``repo_pom_paths``, ``find_parents``. These options behave as follows:

Reviewer 2: I wonder if we want the user to modify the source code of Macaron in ``<macaron_root_directory>/src/config/``?. I think because this section give instructions on how to run Macaron as a Docker image, we instructs the user to provide a custom `defaults.ini` with the relevant field and run with:
```
./run_macaron.sh -dp <path_to_custom_defaults> analyze ...
```

- ``find_repos`` (Values: True or False) - Enables or disables the Repository Finding feature.
- ``artifact_repositories`` (Values: List of URLs) - Determines the remote artifact repositories to attempt to retrieve dependency information from.
- ``repo_pom_paths`` (Values: List of POM tags) - Determines where to search for repository information in the POM files. E.g. scm.url.

> - ``find_parents`` (Values: True or False) - When enabled, the Repository Finding feature will also search for repository URLs in parents POM files of the current dependency.

Reviewer: How about `artifact_ignore_list =`?

requirement. For example, `the Mysql Connector/J <https://slsa.dev/spec/v0.1/requirements#build-service>`_ dependency in
the Micronaut MuShop project does not pass the ``build_service`` check, but can be manually investigated and exempted if trusted. Overall, policies expressed in Datalog can be
enforced by Macaron as part of your CI/CD pipeline to detect regressions or unexpected behavior.

---------------------------
Analyzing more dependencies

> ---------------------------

Reviewer: I think it would be better to move this section as a subsection of `Analyzing with an SBOM`. What do you think?

> In some cases the dependencies that Macaron discovers lack a direct connection to a repository for it to analyze. To improve results in these instances, the Repository Finding feature can be enabled. This feature makes use of a dependency's identifying information and the matching information that can be found within package managing repositories located on the Internet.

Reviewer: ```suggestion

In some cases the dependencies that Macaron discovers lack a direct connection to a repository for it to analyze. To improve results in these instances, the Repository Finding feature can be enabled. This feature makes use of a dependency's identifying information that can be found using the package registries located on the Internet.
```
